src/libopensc/pkcs15-idprime.c

```diff
@@ -0,0 +1,268 @@
+	memset(&serial, 0, sizeof(serial));
+	r = sc_card_ctl(card, SC_CARDCTL_GET_SERIALNR, &serial);
+	if (r < 0) {
+		sc_log(card->ctx, "sc_card_ctl rc=%d", r);
+		p15card->tokeninfo->serial_number = strdup("00000000");
+	} else {
+		sc_bin_to_hex(serial.value, serial.len, buf, sizeof(buf), 0);
+		p15card->tokeninfo->serial_number = strdup(buf);
+	}
+	/* set pin */
+	sc_log(card->ctx,  "IDPrime adding pin...");
+	memset(&pin_info, 0, sizeof(pin_info));
+	memset(&pin_obj,  0, sizeof(pin_obj));
+
+	pin_info.auth_type = SC_PKCS15_PIN_AUTH_TYPE_PIN;
+	sc_pkcs15_format_id(pin_id, &pin_info.auth_id);
+	pin_info.attrs.pin.reference     = 0x11;
+	pin_info.attrs.pin.flags         = SC_PKCS15_PIN_FLAG_INITIALIZED;
+	pin_info.attrs.pin.type          = SC_PKCS15_PIN_TYPE_ASCII_NUMERIC;
+	pin_info.attrs.pin.min_length    = 4;
+	pin_info.attrs.pin.stored_length = 8;
+	pin_info.attrs.pin.max_length    = 8;
+	sc_format_path("", &pin_info.path);
+	pin_info.tries_left    = -1;
+
+	sc_log(card->ctx,  "IDPrime Adding pin with label=%s", pin_label);
+	strncpy(pin_obj.label, pin_label, SC_PKCS15_MAX_LABEL_SIZE - 1);
+	pin_obj.flags = SC_PKCS15_CO_FLAG_PRIVATE;
+
+	r = sc_pkcs15emu_add_pin_obj(p15card, &pin_obj, &pin_info);
+	if (r < 0)
+		LOG_FUNC_RETURN(card->ctx, r);
+
+	/*
+	 * certs, pubkeys and priv keys are related and we assume
+	 * they are in order
+	 * We need to read the cert, get modulus and keylen
+	 * We use those for the pubkey, and priv key objects.
+	 */
+	sc_log(card->ctx,  "IDPrime adding certs, pub and priv keys...");
+	r = (card->ops->card_ctl)(card, SC_CARDCTL_IDPRIME_INIT_GET_OBJECTS, &count);
+	LOG_TEST_RET(card->ctx, r, "Can not initiate cert objects.");
+
+	for (i = 0; i < count; i++) {
+		struct sc_pkcs15_prkey_info prkey_info;
+		struct sc_pkcs15_cert_info cert_info;
+		struct sc_pkcs15_pubkey_info pubkey_info;
+		struct sc_pkcs15_object cert_obj;
+		struct sc_pkcs15_object pubkey_obj;
+		struct sc_pkcs15_object prkey_obj;
+		sc_pkcs15_der_t cert_der;
+		sc_pkcs15_cert_t *cert_out = NULL;
+
+		r = (card->ops->card_ctl)(card, SC_CARDCTL_IDPRIME_GET_NEXT_OBJECT, &prkey_info);
+		LOG_TEST_RET(card->ctx, r, "Can not get next object");
+
+		memset(&cert_info, 0, sizeof(cert_info));
+		memset(&pubkey_info, 0, sizeof(pubkey_info));
+		/* prkey_info cleaned by the card_ctl call */
+		memset(&cert_obj,  0, sizeof(cert_obj));
+		memset(&pubkey_obj,  0, sizeof(pubkey_obj));
+		memset(&prkey_obj,  0, sizeof(prkey_obj));
+
+		cert_info.id = prkey_info.id;
+		pubkey_info.id = prkey_info.id;
+		cert_info.path = prkey_info.path;
+		/* For private keys, we no longer care for the path, just
+		 * the key reference later used in the security environment */
+		prkey_info.path.len = 0;
+		prkey_info.path.aid.len = 0;
+		pubkey_info.key_reference = prkey_info.key_reference;
+		sc_log(card->ctx,  "Key ref r=%x", prkey_info.key_reference);
+
+		pubkey_info.native        = 1;
+		prkey_info.native        = 1;
+
+		memcpy(cert_obj.label, cert_label, sizeof(cert_label));
```

Comment:
These labels are rather useless. Would it make sense to add a counter to the label or to use the CN from the certificate as label?

Reply:
The official driver lists the labels just as some undocumented hashes. I will add a counter since the CN names will show up in the subject field anyway.